test/studies/prk/DGEMM2/dgemm.chpl

```diff
@@ -0,0 +1,115 @@
+use Time;
+use BlockDist;
+use LinearAlgebra;
+
+config type dtype = real;
+
+config param useBlockDist = false;
+
+config const order = 10,
+             epsilon = 1e-8,
+             iterations = 100,
+             windowSize = 10,
+             debug = false,
+             validate = true,
+             correctness = false; // being run in start_test
+
+const vecRange = 0..#order;
+const matrixSpace = {vecRange, vecRange};
+
+const matrixDom = matrixSpace dmapped if useBlockDist then
+                      new dmap(new Block(boundingBox=matrixSpace)) else
+                      defaultDist;
+
+var A = Matrix(matrixDom, dtype),
+    B = Matrix(matrixDom, dtype),
+    C = Matrix(matrixDom, dtype);
+
+forall (i,j) in matrixDom {
+  A[i,j] = j;
+  B[i,j] = j;
+  C[i,j] = 0;
+}
+
+const nTasksPerLocale = here.maxTaskPar;
+
+if !correctness {
+  writeln("Chapel Dense matrix-matrix multiplication");
+  writeln("Max parallelism      =   ", nTasksPerLocale);
+  writeln("Matrix order         =   ", order);
+  writeln("Window size      =   ", if windowSize>0 then 
+      windowSize:string
+      else "N/A");
+  writeln("Number of iterations =   ", iterations);
+  writeln();
+}
+
+const refChecksum = (iterations+1) *
+    (0.25*order*order*order*(order-1.0)*(order-1.0));
+
+var t = new Timer();
+
+private inline iter block(indexes, window : int) {
+  var lowVal = indexes.low;
+  while (true) {
+    var highVal = lowVal + indexes.stride * window - 1;
+    if (highVal >= indexes.high) {
+      yield lowVal..indexes.high;
+      break;
+    } else {
+      yield lowVal..highVal;
+    }
+    lowVal = highVal + 1;
+  }
+}
+
+if windowSize == 0 {
+  for niter in 0..iterations {
+    if niter==1 then t.start();
+
+    forall (i,j) in matrixSpace do
+      for k in vecRange do
+        C[i,j] += A[i,k] * B[k,j];
+
+  }
+  t.stop();
+}
+else {
+
+  coforall loc in A.targetLocales() with (ref t) {
+    on loc {
+      var localDomainA = A.localSubdomain();
+      var localDomainB = B.localSubdomain();
+      var localDomainC = C.localSubdomain();
+
+      const commonDimension = A.domain.dim(1);
+
+      for niter in 0..iterations {
+        if here.id==0 && niter==1 then t.start();
+        for subArrayAChunk in block(A.domain.dim(1), windowSize) {
+          var subArrayA : [localDomainA.dim(0), subArrayAChunk] A.eltType 
+                        = A[localDomainA.dim(0), subArrayAChunk];
+          var subArrayB : [subArrayAChunk, localDomainB.dim(1)] B.eltType 
+                        = B[subArrayAChunk ,localDomainB.dim(1)];
+          C[localDomainC] += dot(subArrayA, subArrayB);
```

Comment:
This is not ideal for performance.

`dot` creates an array, stores the result in it and then returns it. Then we do `+=` over an array slice and the returned array, which is essentially another `forall` after `dot`.

Few things we can do to improve this line:
1. Don't rely on `dot`
2. Discuss whether we can add a new overload of `dot` in the LinearAlgebra module where you give the output array
3. Leave it as-is for now.

I think we should have a better-performing version soon. But OK with not having that in this PR. We can also envision a future where we have this and some other version that circumvents that overhead. It'd be cool to have them both and have an optimization for such cases and see the gap between two versions closing.

Another smaller note is the `localSlice` might be a better way to slice into C: https://chapel-lang.org/docs/master/builtins/ChapelArray.html?highlight=localslice#ChapelArray.localSlice

Reply:
Isn't `subArrayA` and `subArrayB` being fetched synchronously? Also its being fetched entirely by locale 0? One potential improvement: perform blocks in parallel? I'm not even sure if bulk fetching that borders multiple locales is concurrent... is it? I.E. would there be sequential bulk GETs to each locale or is it a multi-locale PUT?

Reply:
Woops, missed the `coforall` above.

Reply:
Using a localSlice improved the performance a bunch